Webhook receivers

jbradberry opened the pull request:
##### SUMMARY
Provide API endpoints that can receive and understand webhook events from some common services we might want to handle.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

jbradberry committed:
Add the webhook-specific fields to JobTemplate and WorkflowJobTemplate

jbradberry committed:
Beginnings of the API views for the webhook receivers

jbradberry committed:
Hook in the webhook receiver views into the urlconf

AlanCoding commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,109 @@
+
+        # Use the hmac `compare_digest` helper function to prevent timing analysis by attackers.
+        if not hmac.compare_digest(force_bytes(obj.webhook_key), self.get_signature()):
+            raise PermissionDenied
```
`()`?

jbradberry replied:
You can raise an Exception class, not just Exception instances.

jbradberry committed:
Forbid access to the webhook receiver views if webhook_key is not set

jbradberry committed:
Move the webhook url include from the top level urlconf to the JT/WFJT urlconfs

jbradberry committed:
Add an api view for obtaining and rotating the webhook key

jbradberry committed:
Attempt to get the RBAC right on the webhook secret key view

jbradberry committed:
Switch to using a permission class for the webhook secret key view

This view is now behaving as expected for superuser, org admin, JT
admin, JT exec, and org member roles.

jbradberry committed:
Add the related link to the webhook secrets view to the serializers

jbradberry committed:
Fix problem with the tests by dynamically setting the view model

instead of using a model @property or lookup method.

jbradberry committed:
Include a check for the webhook_key related resource url

in the tests for JTs and WFJTs.

jbradberry committed:
Revert to using the explicit dispatch to the appropriate model

since passing the model class at url include time doesn't work.

jbradberry committed:
Add some RBAC oriented tests for the webhook secret key view

jbradberry committed:
Expose the webhook_service and webhook_credential fields in the serializer

webhook_credential specifically as a summary field.

AlanCoding commented on awx/main/tests/functional/api/test_webhooks.py:
```diff
@@ -0,0 +1,113 @@
+
+    url = reverse('api:webhook_key', kwargs={'model_kwarg': 'job_templates', 'pk': jt.pk})
+    response = get(url, user=user)
+    assert response.status_code == expect
```
Can we please combine to

```python
response = get(url, user=user, expect=expect)
```

jbradberry replied:
hm, yeah, this is a thing?  Ok.

AlanCoding commented on awx/main/tests/functional/api/test_webhooks.py:
```diff
@@ -0,0 +1,113 @@
+    response = post(url, {}, user=user)
+    assert response.status_code == expect
+    if expect < 400:
+        assert bool(response.data.get('webhook_key'))
```
What's the implication of this assertion? The data you gave was `{}`. Does that mean it auto-generates a secret key?

jbradberry replied:
Yep, that's correct.  We're only going to allow cycling the key, not posting arbitrary user-defined keys.

ryanpetrello replied:
Sounds reasonable to me.

jbradberry committed:
Update tests to use the `expect` keyword argument for get() and post()

jbradberry committed:
Cycle or unset the webhook key if the webhook service changes

Also, tests.

jbradberry committed:
Remove some dead comments

jbradberry committed:
Get the webhook receiver views to work at least minimally

jbradberry committed:
Launch a Job or WorkflowJob based on the incoming webhook

jbradberry committed:
Construct an ID for Gitlab webhooks

by taking the SHA1 of the body of the webhook request.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -118,8 +118,8 @@
-    'job_template': DEFAULT_SUMMARY_FIELDS,
-    'workflow_job_template': DEFAULT_SUMMARY_FIELDS,
+    'job_template': DEFAULT_SUMMARY_FIELDS + ('webhook_credential',),
+    'workflow_job_template': DEFAULT_SUMMARY_FIELDS + ('webhook_credential',),
```
I don't get what this is doing. This is not a list of job template fields which are included in summary_fields for job template. This is a list of fields that will be shown in the JT summarization for the summary_fields of other resources that point to the JT. Since the credential is a related resource of the JT, it would be recursively nesting summary fields to make any sense, or it would be a pk value, which we shouldn't do there.

I would want to know firstly what you're trying to accomplish, and also what the serializer actually gives with this change put in.

jbradberry replied:
Yes, this almost certainly is a mistake on my part.  I hadn't done enough with the credentials portion of this work to have noticed yet.

AlanCoding replied:
it's possible that what you want is a new entry in this list

```python
    'webhook_credential': DEFAULT_SUMMARY_FIELDS,
```

(moving discussion to other comment)

AlanCoding commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +484,48 @@ def get_active_jobs(self):
+        null=True,
+        on_delete=models.SET_NULL,
+        related_name='%(class)ss'
+    )
```
We've just wrapped up a few years of churn involved with adding credential related fields to the JT polymorphic sub-model, and then slowly removing them and deprecating them.

If there's any possible way we could include this in the general `credentials` relationship, which is already present for all unified models, I would greatly prefer that.

For any new ForeignKey on models like JTs and WFJTs, prefetching is not an option. The credentials relationship works around this limitation.

We also already have logic to enforce that it doesn't attach more than 1 of the same type of credential.

jbradberry replied:
I don't know, that strikes me as odd at the very least.  Semantically, these are used for two different things -- `credentials` grants access to machines in the inventory in order to execute tasks, and `webhook_credential` is required to post job status updates to the Github (or other service) API.  cc @ryanpetrello @matburt

AlanCoding replied:
From my other comment, if you're going to want to summarize it as we do for other credentials:

```python
    'webhook_credential': DEFAULT_SUMMARY_FIELDS + ('kind', 'cloud', 'credential_type_id'),
```

The field "kind" comes from a field on the credential type. That means that in order to serialize a job template, both the webhook credential and its credential type will be necessary.

We get both of those for "free" (we are already performing the prefetches necessary) for the credentials relationship. In the location the field is right now, this will unavoidably require at least 1 new query every time a JT/WFJT is serialized that doesn't have a webhook credential, and 2 every time it does.

Since other people are going to be weighing in with judgement calls, I'm just trying to get the facts straight first.

matburt replied:
I'm not sure it's *strictly* necessary to create `generalized` credentials containers *everywhere*. Sometimes we really just need *one* type of credential. I think this is fine here. If we want to end up supporting multiple webhooks per-endpoint then we might feel different but I feel this would have to change in major ways to support that... given that there are so few webhook sources it's probably fine to have a customer support a 1:1 linkage between a webhook source and a job template-ish.

AlanCoding replied:
@matburt should it be shown in JT/WFJT summary_fields?

matburt replied:
Yes, I think so. This credential feels less like the catchall `credentials` since this one won't be included in the playbook runs. I'm open to being conviced otherwise?

ryanpetrello replied:
As much as I like the new `.credentials` M2M, the general purpose of that is intended to be "credentials we use for the job run".  I could see the argument for this being a distinct new foreign key.  What did we do for the Insights credential type and its' relationship?

AlanCoding replied:
> What did we do for the Insights credential type and its' relationship?

That's a foreign key from inventory, `insights_credential`, yes.

Projects are still sporting a field `credential` for their scm credential (from the subclass table), which is most similar to this.

And actually, I think that the null case shouldn't hurt serialization. It will muck hard with copy / start user_capabilities. We will need to return to that before the end of this cycle and make sure it's not doing something crazy we didn't expect.

matburt reviewed (commented):
This is looking good. We need to make sure we add something for the `/docs` architecture documentation before merging.

AlanCoding commented:
More things coming to mind...

Will (or should) a job launched by a webhook have any reference to the webhook credential used? Is the webhook credential always static to the JT / WFJT? Can someone connect a github webhook and a bitbucket webhook to a JT?

If a webhook launches a JT, and then the user modifies the webhook credential on the JT, are there any scenario where that creates a problem?

Can users relaunch jobs that were created by a webhook? Would that create problems?

jbradberry commented:
> More things coming to mind...
> 
> Will (or should) a job launched by a webhook have any reference to the webhook credential used?

This is an extremely good point.  I think that'll depend on how I wind up implementing the "bookend" tasks to do the status posting back to the SCM.

> Is the webhook credential always static to the JT / WFJT? Can someone connect a github webhook and a bitbucket webhook to a JT?

The design we all settled upon was to have a limit to only one configured webhook service per JT / WFJT.

> If a webhook launches a JT, and then the user modifies the webhook credential on the JT, are there any scenario where that creates a problem?

In the case where the SCM does not yet have the new credential, it would mean that the attempts to update the job status on the SCM would fail but presumably the job itself would be fine.  Ditto if the system tasks to send the update has the stale credential but it has been removed from the SCM.

> Can users relaunch jobs that were created by a webhook? Would that create problems?

I don't see why not.  The payload is getting stashed in the extra vars, and the branch or ref that is needed for execution should also be in the appropriate field.  I think the only question, then, is how are the SCM status updating tasks orchestrated and do we want them to still fire off on relaunches.

AlanCoding commented:
Thanks for the replies. There's one area I want to drill a little further down into.

> I don't see why not. The payload is getting stashed in the extra vars, and the branch or ref that is needed for execution should also be in the appropriate field. I think the only question, then, is how are the SCM status updating tasks orchestrated and do we want them to still fire off on relaunches.

If a job was launched by a webhook, will there be an indicator in the UI for this? Is this information that users should know when looking at the jobs list? If a user relaunches a job from a webhook, we may or may not prefer that the UI indicator that it came from a webhook is present for the relaunched job.

jbradberry commented:
> Thanks for the replies. There's one area I want to drill a little further down into.
> 
> > I don't see why not. The payload is getting stashed in the extra vars, and the branch or ref that is needed for execution should also be in the appropriate field. I think the only question, then, is how are the SCM status updating tasks orchestrated and do we want them to still fire off on relaunches.
> 
> If a job was launched by a webhook, will there be an indicator in the UI for this? Is this information that users should know when looking at the jobs list? If a user relaunches a job from a webhook, we may or may not prefer that the UI indicator that it came from a webhook is present for the relaunched job.

Also an excellent point.  Yes, I expect that we would want such an indicator.

jbradberry committed:
Add a new set of personal access token credential types

jbradberry committed:
Fix the summary fields for webhook_credential

jbradberry committed:
Update test_default_cred_types to include the new personal access token types

jbradberry committed:
Validate the webhook credential

- we should allow a null credential, so that the admin can choose to configure not posting back status changes of the triggered job
- the credential must be of the new 'token' kind
- if we do configure a credential, its type must match the selected SCM service

jbradberry committed:
Create a new model mixin for Job and WorkflowJob webhook fields

jbradberry committed:
Debounce when multiple copies of the same webhook event come in

jakemcdermott committed:
add webhook service field

jakemcdermott committed:
add webhook url field

jbradberry committed:
Add the webhook receiver url to the related urls in the serializers

jakemcdermott committed:
add webhook credential field

jakemcdermott committed:
wip

jakemcdermott committed:
issue network calls for setting and getting webhook key

jbradberry committed:
Make sure that the new webhook fields are populated when firing off a job

Also, added a temporary hacky workaround for the fact that something
in our request/response stack for APIView is consuming the request
contents in an unfriendly way, preventing the `.body` @property from
working.

jbradberry committed:
Include a message in the webhook response

jbradberry committed:
Expose the new webhook fields in the job and workflow serializers

jbradberry committed:
Disable the authentication classes for the webhook receivers

One of them was consuming the body of the posts.  We do still need to
have an extraneous `request.body` expression, though now in
WebhookReceiverBase.post, since the `request.data` expression in the
logging also consumes the request body.

jakemcdermott committed:
support server-side webhook key generation

jakemcdermott committed:
use key icon for webhook cred

jakemcdermott committed:
add webhook fields to workflows

jakemcdermott committed:
add launched-by-webhook details to job runs

jakemcdermott committed:
always show launched by webhook details if there's a webhook guid

jakemcdermott committed:
represent webhooks on job lists

jbradberry committed:
Pull out a git ref for each event type where we might care

jakemcdermott committed:
add webhook fields to workflow unit test mock

jakemcdermott committed:
add trailing '/' to webhook urls

jbradberry committed:
Update the webhook receiver git ref extractor logic

to deal with the null-ref case, and to deal correctly with Github push events.

jakemcdermott committed:
stub options request in workflow add unit test

jbradberry committed:
Fix the git ref extractor for Gitlab pull requests

jbradberry committed:
Add logic to post the job status for webhooks back to the service

under some circumstances.

jakemcdermott committed:
make label consistent with help text

jbradberry committed:
Provide a payload for the webhook status post-back

jbradberry committed:
Enable the call to update_webhook_status

by calling it directly within send_notification_templates.  Also,
update the context field in the payload to be either 'ansible/awx' or
'ansible/tower', depending on which is being used.

jbradberry committed:
Posting webhook status now works

jakemcdermott committed:
Always include selected webhook service in creation requests

jbradberry committed:
Drop Bitbucket support

since only the Bitbucket Server product supports signed payloads,
bitbucket.org does not.  And we are requiring signed payloads.

jakemcdermott commented:
recheck

jakemcdermott committed:
Don't alert user of 403 errors for webhook key

jakemcdermott committed:
Apply 403 alert fixes for Workflows, too

jbradberry committed:
Update the Webhook Credential help text tooltip

to make it more apparent to the user that this is an optional part of
the feature, and that failure to add a webhook credential will disable
status post-backs.

jbradberry committed:
Fix problems with posting to Gitlab's API

jbradberry committed:
Fix the webhook receiver url for workflow jobs

kdelee committed:
add arguments to awxkit for webhooks on jt or wfjt

jbradberry committed:
Raise a validation error if a credential is set while the service is not

kdelee commented on awx/main/tests/functional/api/test_webhooks.py:
```diff
@@ -202,3 +202,29 @@ def test_set_wrong_service_webhook_credential(organization_factory, job_template
+@pytest.mark.parametrize(
+    "service", [s for s, _ in WebhookTemplateMixin.SERVICES]
+)
+def test_unset_webhook_service_with_credential(organization_factory, job_template_factory, patch, service):
```
I will try this out. The scenario I was trying to describe was an initial POST to create the job template with only a credential and no service, which worked (although I would have expected it to tell me I needed to set the service)

kdelee replied:
so this is a different scenario (where the JT already existed with a service and a credential, but then the service is removed)

jbradberry replied:
Ok, I added another test a bit further above that captures this case.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2825,6 +2826,25 @@ def get_summary_fields(self, obj):
+        webhook_service = attrs.get('webhook_service', getattr(self.instance, 'webhook_service', None))
+        webhook_credential = attrs.get('webhook_credential', getattr(self.instance, 'webhook_credential', None))
+
+        if webhook_credential and webhook_credential.credential_type.kind != 'token':
```
`kind` has been the target of deprecation attempts

I would prefer `webhook_credential.credential_type.namespace`

jbradberry replied:
All right, how about now?

jbradberry committed:
Avoid the use of CredentialType.kind

AlanCoding commented:
I would really prefer a documentation template for the net-new pages. Notice how there's no ❓ in the API browser. These are also used for auto-generated docs online.

![Screen Shot 2019-09-26 at 3 49 16 PM](https://user-images.githubusercontent.com/1385596/65720157-631ab180-e075-11e9-84a5-06e16c614934.png)
![Screen Shot 2019-09-26 at 3 49 23 PM](https://user-images.githubusercontent.com/1385596/65720158-631ab180-e075-11e9-8479-ae9dd9f19525.png)

AlanCoding commented:
Diving deeper into OPTIONS, I see that `/api/v2/job_templates/38/webhook_key/` lists no fields, and `/api/v2/job_templates/38/github/` lists no HTTP methods. I guess maybe you could say it doesn't make sense for the latter, since that endpoint cannot or should not be used by the CLI. ~But the webhook_key should definitely provide a notion of what fields could be used~.

On second thought, if OPTIONS doesn't work for `/github` then I think a 405 would be best, right?

Never mind about the webhook_key, no fields are taken, 👍 to options there

AlanCoding commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,242 @@
+        if model is None:
+            raise PermissionDenied
+
+        return model.objects.filter(webhook_service=self.service).exclude(webhook_key='')
```
the field `webhook_key` isn't indexed, right? It might be better to not do the `.exclude` and just raise a not-found Django error if it's blank. But I guess lookups will have to happen at some point anyway.

/minor

jbradberry replied:
Yeah, it's unindexed.  It's unclear whether it's worthwhile to add a potentially large index for it.

ryanpetrello replied:
I don't think an index is particularly important here; we're talking about filtering on the JT and WFJT tables, which aren't particularly large (in comparison to, say, UJ and the event tables).

If we can accomplish the same thing in Python (as @AlanCoding suggested), that's fine, too.  I don't have strong opinions.

AlanCoding commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,242 @@
+        logger.debug(
+            "headers: {}\n"
+            "data: {}\n".format(request.headers, request.data)
+        )
```
Is this log something that should be deleted before merge??

jbradberry replied:
No, it was explicitly discussed several times in the design meetings.  Or, at least, this is my interpretation of such.

ryanpetrello replied:
👍

jbradberry committed:
Add a new launch_type of 'webhook'

AlanCoding commented:
Implementation of the new launch type looks good 👍  thanks

AlanCoding commented:
I think the code looks fine to me so far, but I don't see public docs along with this that describe the use cases in the final form of the implementation.

Working backwards, I know that there's a callback at the finish of a job to the service that launched it. This is probably interpreted as a CI check? I'd imagine that probably won't apply in all situations, and that there's a lot of details along with that.

I also just realized I need to go back over the SCM branch handling in greater details.

jakemcdermott committed:
Use job launch_type field to detect webhook jobs

We have a launch type field for categorizing the different ways jobs
can be launched. This updates the UI to use this field when checking
if a job was launched by a webhook.

AlanCoding commented on awx/ui/client/features/jobs/jobsList.controller.js:
```diff
@@ -159,7 +159,7 @@ function ListJobsController (
 
     vm.getWebhookDetails = (job) => {
-        if (!job.webhook_guid) {
+        if (job.launch_type !== 'webhook') {
```
Nice. So I assume that the heuristic `!job.webhook_guid` would still work. Switching over to launch_type just makes it cleaner.

jbradberry replied:
Yep

jbradberry committed:
Add a doc template to the webhook key API view

jbradberry commented:
> I think the code looks fine to me so far, but I don't see public docs along with this that describe the use cases in the final form of the implementation.
> 
> Working backwards, I know that there's a callback at the finish of a job to the service that launched it. This is probably interpreted as a CI check? I'd imagine that probably won't apply in all situations, and that there's a lot of details along with that.

Yes, that's correct, but the callback is optional.  You can leave it unconfigured by not adding a `webhook_credential`.

AlanCoding commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +489,132 @@ def get_active_jobs(self):
+    webhook_service = models.CharField(
+        max_length=16,
+        choices=SERVICES,
+        blank=True
```
help_text for these 3 new fields would help to document what this does.

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -975,6 +976,38 @@ def create(self):
+    inputs={
+        'fields': [{
+            'id': 'token',
+            'label': ugettext_noop('Token'),
```
it's also possible to [add help_text to fields](https://github.com/ansible/awx/blob/devel/docs/credentials/custom_credential_types.md) here, and that might also help. Presumably this comes from somewhere in the github repo settings??

jbradberry committed:
Add help_text to the new fields

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2825,6 +2826,25 @@ def get_summary_fields(self, obj):
+        webhook_credential = attrs.get('webhook_credential', getattr(self.instance, 'webhook_credential', None))
+
+        if webhook_credential:
+            if not webhook_credential.credential_type.namespace.endswith('_token'):
```
I think @AlanCoding may have commented on this already and suggested something like this, but I think that `kind == 'token'` *is* more appropriate here.

https://github.com/ansible/awx/pull/4436/files#diff-8923e305beef7163bf0fcb6d87998376R991

`CredentialType`s *do* have a  `kind`; `Credential.kind` is the type of thing we've been trying to shy away from.  @AlanCoding thoughts?

jbradberry replied:
Yeah, that would be undoing exactly what Alan suggested.

ryanpetrello replied:
Alright, @AlanCoding, defend your thesis.

AlanCoding replied:
Sorry, I believe I did misunderstand the function at the time. credential type kind is like `scm`, `vault`, `ssh`, `insights`, and the kind here is `token`...

I need to consult on words to use for things. This is a new category of credential, that does make sense to me. Its function is to update the status of CI checks (these do not feel like good words). Maybe `ct.kind == 'webhook'` would be readable.

ryanpetrello commented on awx/api/templates/api/webhook_key_view.md:
```diff
@@ -0,0 +1,11 @@
+  webhook configuration of the service this template will be receiving
+  webhook events from (string, read-only)
+
+Make an empty POST request to this resource to cycle in a new `webhook_key`.
```
"to cycle in a new ..."

maybe:

"to regenerate a new replacement webhook_key"

ryanpetrello commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+        }
+        if WorkflowJob.objects.filter(**kwargs).exists() or Job.objects.filter(**kwargs).exists():
+            # Short circuit if this webhook has already been received and acted upon.
+            logger.debug("Webhook previously received, returning without action.")
```
👍

ryanpetrello commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+        if WorkflowJob.objects.filter(**kwargs).exists() or Job.objects.filter(**kwargs).exists():
+            # Short circuit if this webhook has already been received and acted upon.
+            logger.debug("Webhook previously received, returning without action.")
+            return Response({'message': "Webhook previously received, aborting."},
```
We should i18n this value w/ `_()`

ryanpetrello commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+        return Response({'message': "Job queued."}, status=status.HTTP_202_ACCEPTED)
+
+
+class GithubWebhookReceiver(WebhookReceiverBase):
```
I really like how this interface turned out here, especially if we find other services we want to support in the future 👍

ryanpetrello commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+
+    def get_signature(self):
+        header_sig = self.request.META.get('HTTP_X_HUB_SIGNATURE')
+        if not header_sig:
```
Might be good to add some `logger.debug` here...

jbradberry replied:
Done.

ryanpetrello commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+            raise PermissionDenied
+        hash_alg, signature = header_sig.split('=')
+        if hash_alg != 'sha1':
+            raise PermissionDenied
```
...and here, to give us some clues in case APIs change at some point.

jbradberry replied:
Done.

ryanpetrello commented on awx/main/migrations/0091_v360_webhook_mixin.py:
```diff
@@ -0,0 +1,49 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('main', '0090_v360_WFJT_prompts'),
```
You and @beeankha are gonna compete for `0091`, and I suspect she'll beat you to the merge button.

jbradberry replied:
That's ok, I have my script to re-number and re-dependency migrations during a rebase.

ryanpetrello commented on awx/main/models/credential/__init__.py:
```diff
@@ -968,6 +969,40 @@ def create(self):
+            'label': ugettext_noop('Token'),
+            'type': 'string',
+            'secret': True,
+            'help_text': ugettext_noop('This token needs to come from your profile settings in Github')
```
👍

ryanpetrello commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +489,138 @@ def get_active_jobs(self):
+        abstract = True
+
+    SERVICES = [
+        ('github', "Github"),
```
I think technically speaking, the names are GitHub and GitLab

jbradberry replied:
So it is.  How ugly.

ryanpetrello commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +489,138 @@ def get_active_jobs(self):
+                'pending': 'pending',
+                'successful': 'success',
+                'failed': 'failure',
+                'canceled': 'failure',  # Github doesn't have a 'canceled' status :(
```
👎

ryanpetrello commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +489,138 @@ def get_active_jobs(self):
+            license_type = get_licenser().validate().get('license_type')
+            data = {
+                'state': statuses[status],
+                'context': 'ansible/awx' if license_type == 'open' else 'ansible/tower',
```
This was a neat idea.

ryanpetrello commented on awx/main/models/mixins.py:
```diff
@@ -483,3 +489,138 @@ def get_active_jobs(self):
+                k: v.format(self.webhook_credential.get_input('token')),
+                'Content-Type': 'application/json'
+            }
+            response = requests.post(status_api, data=json.dumps(data), headers=headers)
```
We may want to specify a sane default `timeout=` (I can never remember if requests does this automatically or not).

jbradberry replied:
It does not.  Adding...

AlanCoding commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+    @csrf_exempt
+    def post(self, request, *args, **kwargs):
+        # Ensure that the full contents of the request are captured for multiple uses.
+        request.body
```
The line `request.body` does

> Ensure that the full contents of the request are captured for multiple uses

?

jbradberry replied:
So this was annoying af.  Something upstream in the request/response cycle was consuming data out of the request without using the `request.body` property (which has a condition to cache the full contents onto an attribute on the request, so that subsequent uses of this property also have access).  This made it so that I couldn't use `request.body` at all without raising an exception, since all of the data had been read but not stored somewhere.

So, I manipulated `APIView` a bit to take care of something that was causing it there, but since my logging call right below this does something similar I've got to kind of prime the pump here.  If that makes sense?

AlanCoding replied:
oh, it really is as nasty as it looks

thanks for explaining that.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -1206,6 +1207,8 @@ def _websocket_emit_status(self, status):
 
     def websocket_emit_status(self, status):
         connection.on_commit(lambda: self._websocket_emit_status(status))
+        if hasattr(self, 'update_webhook_status'):
```
I like this.

ryanpetrello reviewed (commented):
This is looking really good, just some minor feedback.

AlanCoding commented on awx/api/views/webhooks.py:
```diff
@@ -0,0 +1,243 @@
+            'webhook_service': obj.webhook_service,
+            'webhook_guid': event_guid,
+        }
+        if WorkflowJob.objects.filter(**kwargs).exists() or Job.objects.filter(**kwargs).exists():
```
If `webhook_guid` is a unique identifier for the event that triggered the job, then what does that correspond to? A pull request? Any modification to a pull request?

Wouldn't it be plausible for a user to configure multiple JTs / WFJTs to run when a single pull request was received. Maybe they trigger independent checks? That may or may not be inconsistent with this code, it depends on the exact meaning of the guid.

jbradberry replied:
Oh.  :(

jbradberry replied:
In the PR context, it would correspond to a specific push (with the caveat that a push to a PR triggers like 3 different events, each with their own GUID: an update to a PR, a push, and ... a change to a ref? maybe)

jbradberry replied:
You are absolutely correct, though.  Great catch.  I guess what we _actually_ want is to check the GUID + the template id.

jbradberry replied:
Should be fixed now.

jbradberry committed:
Address a variety of small review issues

jbradberry committed:
Modify the webhook debounce logic

to check if we've already previously run a job with the same webhook
GUID plus template id.  This will allow organizations to write
multiple JT/WFJTs to handle the same set of webhook events.

ryanpetrello reviewed (approved):
Unless @AlanCoding has any other notable feedback, I'm pretty happy with this 👍

jbradberry committed:
Return to using ContentType.kind

which is _not_ the `kind` attribute being deprecated.

jbradberry committed:
Post the job or workflow url to the webhook service as part of the status

AlanCoding reviewed (approved):
The API code looks solid, I'm fine to merge this if we can apply a needs_docs label. Even for you to just enumerate the manual steps you took and the outcomes expected, that would be enough. The `docs/` folder would be good for this.

Thanks for renaming the CT kind and adding the launch_type.

jbradberry committed:
Change the webhook post-back payload to use the job UI url

jlmitch5 commented on awx/ui/client/features/jobs/jobsList.controller.js:
```diff
@@ -158,6 +158,14 @@ function ListJobsController (
+            return null;
+        }
+
+        return strings.get('list.ROW_ITEM_LABEL_WEBHOOK');
```
you are returning the word "Webhook"?

jakemcdermott replied:
That's right. This is to show a tag that says `Webhook` on job list items that were launched by a webhook. 

Regarding the UX, I confirmed with @trahman73  a few weeks ago that this is how we want to indicate webhook jobs on the job lists.

As far as the implementation is concerned, this is the same pattern we have in place on this controller for setting the value of the job slice badges, so I figured it was alright: https://github.com/ansible/awx/blob/devel/awx/ui/client/features/jobs/jobsList.controller.js#L155

Did you have a specific change in mind that you'd like to see for this part of the code? I'm open to  suggestions if you have them.

jakemcdermott replied:
should be fixed here: https://github.com/ansible/awx/pull/4436/commits/b4b2cf76f6045ccde8ecf002a6267cf5ea993bdd

jlmitch5 commented on awx/ui/client/src/templates/workflows/edit-workflow/workflow-edit.controller.js:
```diff
@@ -63,6 +63,179 @@ export default [
+                });
+        };
+
+        $('#content-container').append($compile(`
```
this feels a bit strange.  could we go ahead and have this compiled (as part of the view's partial, and then just conditionally show/hide it when some portion of the data related to $scope.webhookCredential is updated?

jakemcdermott replied:
I know what you mean :frowning_face:. It's not ideal, but we might have to stick to doing it this way. Compiling the template inside the add/edit controllers certainly _is_ a hack. I don't know of a better way to get the compiled modal and lookup template binded to the views driven by the add/edit controllers, which is where I need them to be. I'm open to any specific suggestions you might have, otherwise I'm inclined to leave things as they are.

Although the _way_ we get this template compiled onto the view isn't great, the add/edit controllers are the right _place_ for these components given the state they must encapsulate. Appending it to the details partial like you suggest would have been my first choice but the entire view seems to be dynamically generated by `stateDefinitions.factory.js` and `formGenerator.js` and so there's no partial at the detail or form level where one could add the lookup modal component in a more conventional way. :man_shrugging:  

I considered and tried a few alternatives before settling on this as the least bad choice. Here are some of the alternatives to `compile->append` that I considered a few weeks back:

- _Removing_ the template and adding even more code to `formGenerator.js` and `stateDefinitionsFactory.js` to change how they generate their html in the hopes of coercing it to handle all the things I need the new fields to do might work but it would be a risky and unmaintainable solution.

- Moving the template _lower_ into the webhook credential input partial would be very problematic as the behavior of the webhook credential lookup depends on the current (unsaved) value of the webhook service input, which is part of the form and not encapsulated by the webhook credential input component.

- Moving the template _higher_ into the partial of a parent component would mean **either**:
  - appending the webhook lookup component to `features/templates/index.html`, which is confusing as no other part of the details view or form state is managed there.

  - **or**, moving it to `client/index.template.ejs`, which would be really unexpected and probably buggy since that encapsulates nearly the entire application independent of routing.


FWIW, we compile and append [the job template multi-credential lookup in the same way](https://github.com/ansible/awx/commit/b5bc387d335c45af6358723122b20601753b97ad#diff-b3d385ba0eea4499cdd8496c1b3e4a79R69). To my knowledge, this hasn't caused any problems or unexpected behavior to date.

jlmitch5 replied:
ah totally understand, basically I thought that there was a template to put it in, but it looks like there's not.  No problem!

jlmitch5 commented on awx/ui/client/src/templates/workflows/add-workflow/workflow-add.controller.js:
```diff
@@ -52,6 +52,138 @@ export default [
+
+        $scope.handleWebhookKeyButtonClick = () => {};
+
+        $('#content-container').append($compile(`
```
same thing...is there a way to not use $compile?

jakemcdermott replied:
I looked for alternatives, but didn't come up with anything. see comment above: https://github.com/ansible/awx/pull/4436#discussion_r329784135.

jlmitch5 reviewed (changes requested):
code looks good, just a couple of questions that don't quite make sense (may just be me not understanding)

jbradberry committed:
Check for the existance of a UnifiedJobTemplate with the same webhook GUID

instead of trying (incorrectly) to be specific about the JT/WFJT type.

jakemcdermott committed:
Refactor job secondary label assignment

jlmitch5 reviewed (approved):
looks good!
